### PR TITLE
Remove both StorageLive and StorageDead in CopyProp.

### LINKED ISSUE #107511: Bad MIR: StorageLive on a local that was already live with opt-level=4
Consider this code:
```rust
fn main() {
    let mut sum = 0;
    let a = [0, 10, 20, 30];
    for i in 0..a.len() {
        sum += a[i];
    }
    println!("{sum}");
}
```
When running this in Miri with `-Zmir-opt-level=4 -O`, we get an error:
```
error: Undefined Behavior: StorageLive on a local that was already live
 --> tests/pass/loops.rs:4:9
  |
4 |     for i in 0..a.len() {
  |         ^ StorageLive on a local that was already live
  |
  = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
  = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
  = note: BACKTRACE:
  = note: inside `main` at tests/pass/loops.rs:4:9: 4:10
```
This is a fairly recent regression (ad48c109815a2e9441a7ad7796e55b8771fe01a5 was good, a322848c6b0e037c1f0209387558ecb6ab763714 is bad; commit range: https://github.com/rust-lang/rust/compare/ad48c109815a2e9441a7ad7796e55b8771fe01a5...a322848c6b0e037c1f0209387558ecb6ab763714).

Cc @rust-lang/wg-mir-opt

### PR DESCRIPTION
**Pull request by cjgillot**
Fixes https://github.com/rust-lang/rust/issues/107511

https://github.com/rust-lang/rust/pull/106908 removed StorageDead without the accompanying StorageLive. In loops, execution would see repeated StorageLive, without any StorageDead, which is UB.

So when removing storage statements, we have to remove both StorageLive and StorageDead.

~I also added a MIR validation pass for StorageLive. It may be a bit overzealous.~

### CODE REVIEW

**Comment by RalfJung** [comment]
Thanks for the quick fix!

> https://github.com/rust-lang/rust/pull/106908 removed StorageDead without the accompanying StorageLive. In loops, execution would see repeated StorageLive, without any StorageDead, which is UB.

Ah, I see. Makes sense.
(FWIW whether we really want redundant StorageLive to be UB is kind of an open question, but for now it's an invariant we want to enforce -- see https://github.com/rust-lang/rust/issues/99160. If having to remove StorageLive here is for some reason a burden or disadvantage, please leave a note in that issue. It would be nice to enforce the same for StorageDead but due to https://github.com/rust-lang/rust/issues/98896 that is not currently possible.)

**Comment by RalfJung** [comment]
> I also added a MIR validation pass for StorageLive. It may be a bit overzealous.

I'm afraid I can't review that part, I don't know the APIs it uses so I don't even know what it actually checks.

So either r=me with that part moved to a separate PR, or we wait for another reviewer.

File: tests/mir-opt/copy-prop/issue_107511.rs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    let mut sum = 0;
+    let a = [0, 10, 20, 30];
+
+    // `i` is assigned in a loop. Only removing its `StorageDead` would mean that
+    // execution sees repeated `StorageLive`. This would be UB.
+    for i in 0..a.len() {
+        sum += a[i];
+    }
+
+    let _ = sum;
```
Comment (RalfJung) [inline tests/mir-opt/copy-prop/issue_107511.rs:14]: `let _val = sum` might be better, `_` patterns can be quite surprising...

**Comment by cjgillot** [comment]
I don't really have an opinion on whether this should be UB. This is not a (currently) difficult invariant to preserve, so optimizations should probably preserve it.

**Pull request merged**